sdk/storage/azure-storage-blob-changefeed/src/main/java/com/azure/storage/blob/changefeed/Changefeed.java

```diff
@@ -0,0 +1,138 @@
+import com.azure.storage.blob.changefeed.implementation.models.ChangefeedCursor;
+import com.azure.storage.blob.changefeed.implementation.util.DownloadUtils;
+import com.azure.storage.blob.changefeed.implementation.util.TimeUtils;
+import com.azure.storage.blob.models.BlobItem;
+import com.azure.storage.blob.models.ListBlobsOptions;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+import java.io.IOException;
+import java.time.OffsetDateTime;
+
+/**
+ * A class that represents a Changefeed.
+ *
+ * The changefeed is a log of changes that are organized into hourly segments.
+ * The listing of the $blobchangefeed/idx/segments/ virtual directory shows these segments ordered by time.
+ * The path of the segment describes the start of the hourly time-range that the segment represents.
+ * This list can be used to filter out the segments of logs that are interest.
+ *
+ * Note: The time represented by the segment is approximate with bounds of 15 minutes. So to ensure consumption of
+ * all records within a specified time, consume the consecutive previous and next hour segment.
+ */
+class Changefeed {
+
+    private static final String SEGMENT_PREFIX = "idx/segments/";
+    private static final String METADATA_SEGMENT_PATH = "meta/segments.json";
+
+    private final BlobContainerAsyncClient client; /* Changefeed container */
+    private final OffsetDateTime startTime; /* User provided start time. */
+    private final OffsetDateTime endTime; /* User provided end time. */
+    private OffsetDateTime lastConsumable; /* Last consumable time. The latest time the changefeed can safely be
+                                              read from.*/
+    private OffsetDateTime safeEndTime; /* Soonest time between lastConsumable and endTime. */
+    private final ChangefeedCursor cfCursor; /* Cursor associated with changefeed. */
+    private final ChangefeedCursor userCursor; /* User provided cursor. */
+    private final SegmentFactory segmentFactory; /* Segment factory. */
+
+    /**
+     * Creates a new Changefeed.
+     */
+    Changefeed(BlobContainerAsyncClient client, OffsetDateTime startTime, OffsetDateTime endTime,
+        ChangefeedCursor userCursor, SegmentFactory segmentFactory) {
+        this.client = client;
+        this.startTime = startTime;
+        this.endTime = endTime;
+        this.userCursor = userCursor;
+        this.segmentFactory = segmentFactory;
+
+        this.cfCursor = new ChangefeedCursor(this.endTime);
+        this.safeEndTime = endTime;
+    }
+
+    /**
+     * Get all the events for the Changefeed.
+     * @return A reactive stream of {@link BlobChangefeedEventWrapper}
+     */
+    Flux<BlobChangefeedEventWrapper> getEvents() {
+        return validateChangefeed()
+            .flatMap(ignore -> populateLastConsumable())
+            .flatMapMany(ignore -> listYears())
```

Comment:
Given these values are ignored, could we use `then` and `thenMany`?

Reply:
Originally they werent playing well with my error logic but it seems to work now.